Implement new components for phone verification process

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Part of #962.

- Implement a new component `<VerificationCodeControl>` for inputting verification code
- Implement a new component `<VerifyPhoneNumberContent>` for phone verification process
- Add a temporary demo to the Settings page.

### Additional notes

- [x] 📝  The `<DemoVerifyPhoneNumberCard>` should be removed before merging this PR.
- 💡 This PR added new components for subsequence use. It will not change any implementation of the current lite contact info.

### Screenshots:

#### 🎥 . New components for easily entering verification codes

https://user-images.githubusercontent.com/17420811/131971354-5ddfe145-7158-4ecb-bde1-84f40012b9e7.mp4

- Auto-focus to next `<input>` node after entering a valid digit
- Use the arrow keys to move the input cursor
- When key down on the backspace,
    - remove the current digit if the cursor is at the right of digit.
    - move the previous `<input>` if the cursor is at the beginning. And if there's a digit in the previous `<input>`, it would be removed.
- Replace existed digit with the new inputting one

#### 🎥 . The phone verification process

https://user-images.githubusercontent.com/17420811/131969645-6e768b22-903a-4a74-ab75-456ae73debbb.mp4

- Switch verification method and cooldown of the resend button
- The cooldown countdown of the two methods are independent of each other
- Show error message when the verification is failed
- A callback would be invoked when verification is successful

#### 🎥 . Request verification code again

https://user-images.githubusercontent.com/17420811/131970510-c72ca56e-c1da-4333-a22f-801bcc3a971b.mp4

- The verification code request of the same method would be trigger only once when switching the method
- Use the "Resend code" or "Call again" to request code again, and all `<input>` of verification code would be reset to empty. (They would also be reset when switching method but I forgot to record it in. 😅 )
- \*To facilitate the recording of this demo, I changed the cooldown time to 5 seconds locally.

### Detailed test instructions:

1. Go to the Setting page and there's a demo of the two new components for testing
2. Try the UI to see if they work properly and whether it's intuitive and easy to use by entering the verification code.
3. Enter the verification code `000000` to pass the verification, and the rest will fail.

### Changelog entry

eason9487 committed:
Implement a new component for inputting verification code.

eason9487 committed:
Implement the the basic UI of a new card content component for verifying phone number

eason9487 committed:
Add the cooldown timer for locking the next verification code request

eason9487 committed:
Add mock API calls and a prop for verified callback

eason9487 committed:
Add parts of error handling and UI

eason9487 committed:
Add a temporary component to demonstrate the phone number verification card.

eason9487 committed:
Change the returning data type of useCountdown from Array to Object

eason9487 committed:
Add JSDoc for useCountdown hook

eason9487 committed:
Add unit test for useCountdown hook

ecgan commented on js/src/components/contact-information/phone-number-card/useCountdown.js:
```diff
@@ -0,0 +1,76 @@
+	const timerRef = useRef( {} );
+	const [ second, setSecond ] = useState( 0 );
+
+	timerRef.current.usingHandle = handle;
+	timerRef.current[ handle ] = timerRef.current[ handle ] || { callCount: 0 };
```
💅 I think using `handle` and storing the timers inside `timerRef.current[ handle ]` is making things more complicated. It would be simpler to make `useCountdown` deal with just one timer.

eason9487 replied:
I implemented it in this way at first, but as the internal state of the `<VerifyPhoneNumberContent>` component became more and more complex, I tried to move the `handle` (method) here to reduce a little complexity of `<VerifyPhoneNumberContent>`.

I thought it looks elegant when using the `useCountdown` hook with `handle`, though there's a little more complexity encapsulated inside.

ecgan replied:
> I thought it looks elegant when using the `useCountdown` hook with `handle`

Another problem I thought about is that in case we want to use only one fixed timer, then we would need to do it like:

```
const timer = useCountdown('one')
```

Technically we could still do `useCountdown()` (with `undefined`) but then things would start to get a bit weird in the implementation. 😅 

Anyway, maybe this is something we can think about and refactor later; if it works now, then we are good. 

> though there's a little more complexity encapsulated inside.

Yeah. 😅

ecgan commented on js/src/components/contact-information/phone-number-card/useCountdown.js:
```diff
@@ -0,0 +1,76 @@
+ * @param {string} handle The countdown handle.
+ * @return {Countdown} The corresponding countdown states and its start function.
+ */
+export default function useCountdown( handle ) {
```
❔ Not sure if you are aware about this, might be helpful: https://react-hooks.org/docs/usecountdown/

Think we can consider using that library and we don't need to reinvent the wheel and don't need to come out with our tests.

eason9487 replied:
Thanks! I didn't know it. After looking into its doc, I think it's not so suitable for our use case. If we apply it, we have to control the `endTime` state and record the calling count of the countdown timer from outside, or still need to create a custom hook to wrap additional states with that `useCountdown` hook. Consider the implementation of the current `useCountdown` is short and simple, I think we can implement it instead of introducing a new dependency.

ecgan commented on js/src/components/contact-information/phone-number-card/verify-phone-number-content.js:
```diff
@@ -0,0 +1,236 @@
+	onPhoneNumberVerified,
+} ) {
+	const [ method, setMethod ] = useState( verificationMethod );
+	const { second, callCount, startCountdown } = useCountdown( method );
```
💅  Related to my other comment, if we make `useCountdown` deal with one timer, then we don't need to pass in `method`, and we can have something like this:

```js
	const smsCountdown = useCountdown();
	const phoneCountdown = useCountdown();
```

We can do this because there are only two fixed countdowns, the number of countdowns is not dynamic.

ecgan commented on js/src/components/contact-information/phone-number-card/useCountdown.js:
```diff
@@ -0,0 +1,76 @@
+				}
+			};
+
+			timer.id = setInterval( timer.updateSecond, 1000 );
```
TIL and just sharing, I happened to find this online, which is recommending `setTimeout` over `setInterval`: https://stackoverflow.com/a/6685414/379641 

I think `setInterval` for our use here should be okay, don't think we need to worry too much about it.

eason9487 replied:
Indeed, it always needs to be careful when using either `setTimeout` or `setInterval`. 😆

ecgan commented on js/src/components/contact-information/phone-number-card/useCountdown.js:
```diff
@@ -0,0 +1,76 @@
+			}
+
+			timer.updateSecond = () => {
+				let remainingSecond = ( timeUp - new Date().getTime() ) / 1000;
+				remainingSecond = Math.max( Math.round( remainingSecond ), 0 );
```
❔ The time calculation feels a bit like acrobatic (new Date(), Math.max, Math.round), wondering maybe we can make it simpler by manipulating integer second instead, i.e. `60`, `59`, `58`... just minus `1` every one second interval until it becomes `0`.

eason9487 replied:
There're two reasons to calculate it by time factor:

1. I'm not sure if the cooldown time would be a hard rule limited by our or Google's API. If it's limited by API, then it'd be proper to count by time diff.
2. When the process of event loop is blocked, the timer would be blocked as well. Counting by time diff can still result in a more accurate countdown. For example that the timer is blocked by `alert()`:
   ![Kapture 2021-09-07 at 14 39 50](https://user-images.githubusercontent.com/17420811/132296564-bc179a60-721f-4835-ac90-15b2573fb12d.gif)

In most cases, counting by time diff as default preference would not bring surprise when using timers.

ecgan replied:
Thanks for the explanation, I gave some more thoughts about it and yeah it makes sense to use time diff 🙂  👍

ecgan reviewed (commented):
Didn't get to test and review everything, just reviewed a few files and the `useCountdown` timer part.

tomalec commented on js/src/components/contact-information/phone-number-card/constants.js:
```diff
@@ -0,0 +1,2 @@
+export const VERIFICATION_METHOD_SMS = 'SMS';
+export const VERIFICATION_METHOD_PHONE_CALL = 'PHONE_CALL';
```
:nail_care: :nail_care: 
WDYT of changing it to enum? so the person importing the constant, would be sure, they imported everything.
```suggestion
/**
 * …
 *
 * @enum {string}
 */
export const VERIFICATION_METHOD = Object.freeze( {
	SMS: 'SMS',
	PHONE_CALL: 'PHONE_CALL',
} );
```
Also, to emphisize the difference between constant name ('VERIFICATION_METHOD') and values available ('SMS', 'PHONE_CALL')

ecgan replied:
This is nice, I like `VERIFICATION_METHOD.SMS` and `VERIFICATION_METHOD.PHONE_CALL` 🙂  👍

eason9487 replied:
Thanks. I did get stuck on their naming for a while before. Your suggestion was one of my candidates for naming and make me more confident to take it. Changed by https://github.com/woocommerce/google-listings-and-ads/pull/993/commits/57e09b28a8b4359979aea0fd4d0c4b44545cbf38

tomalec commented on js/src/components/contact-information/phone-number-card/useCountdown.test.js:
```diff
@@ -0,0 +1,144 @@
+	it( 'initially should return `{ second: 0, callCount: 0, startCountdown: Function }`', () => {
+		const { result } = renderHook( () => useCountdown( 'A' ) );
+
+		expect( result.current.second ).toBe( 0 );
```
:nail_care: 
Personally, I like to use [`toHaveProperty` matchers](https://jestjs.io/docs/expect#tohavepropertykeypath-value) in such cases to get more descriptive failure messages if somethin gets off
```suggestion
		expect( result.current ).toHaveProperty( 'second', 0 );
```
but `result.current.callCount` is not a complicated expression, and easy to get what is what, so I'm totally fine with `expect( result.current.second ).toBe( 0 );` as it is short.

ecgan replied:
💅  Personally I would prefer to use static dot notation `.second` here, especially when `second` is expected to be always there. I'd use `toHaveProperty` for properties that would be dynamically added and removed at run time.

tomalec replied:
But we are testing here `useCountdown` and that's `useCountdown` that adds `.second`, so without the tested function, it's not always there.

Now, consider someone, unaware that `second` is expected by the outside world, changes it internally to `.remainingSeconds` thinking it's just an implementation detail.

Then, we would get 
```
  ● useCountdown › initially should return `{ second: 0, callCount: 0, startCountdown: Function }`

    expect(received).toBe(expected) // Object.is equality

    Expected: 0
    Received: undefined
```

instead of 
```
  ● useCountdown › initially should return `{ second: 0, callCount: 0, startCountdown: Function }`

    expect(received).toHaveProperty(path, value)

    Expected path: "second"
    Received path: []

    Expected value: 0
    Received value: {"callCount": 0, "remainingSeconds": 0, "startCountdown": [Function anonymous]}

```

I agree, it's not a big difference. The real thing would be if `current` would change 
```
  ● useCountdown › initially should return `{ second: 0, callCount: 0, startCountdown: Function }`

    TypeError: Cannot read property 'second' of undefined
```
(but that's somewhat assured by `useRef` => `React`)

eason9487 replied:
I think the `current` is assured by the `renderHook`. We don't have to worry whether it would be changed.

Not sure if you get the same output as I do, the place below of it indicates which assertion is failed by `>` and `^` symbols, and the content should be descriptive enough.

```
  ● useCountdown › initially should return `{ second: 0, callCount: 0, startCountdown: Function }`

    expect(received).toBe(expected) // Object.is equality

    Expected: 0
    Received: undefined

      13 | 		const { result } = renderHook( () => useCountdown( 'A' ) );
      14 |
    > 15 | 		expect( result.current.second ).toBe( 0 );
         | 		                                ^
      16 | 		expect( result.current.callCount ).toBe( 0 );
```

tomalec commented on js/src/components/contact-information/phone-number-card/useCountdown.test.js:
```diff
@@ -0,0 +1,144 @@
+			jest.useRealTimers();
+		} );
+
+		it( 'should start countdown 10 to 0 by calling `startCountdown( 10 )` and update the current remaining countdown to `second` value after every second.', () => {
```
:nail_care: :grey_question: 
I'm not an English expert, but shouldn't it be more like:
```suggestion
		it( 'should start countdown 10 to 0 by calling `startCountdown( 10 )` and update the `second` property every second with the current remaining countdown.', () => {
```

tomalec commented:
:question: 
I noticed that the countdown takes ~1MB of memory, which looks quite expensive for such a simple component.
![heapsize 1mb](https://user-images.githubusercontent.com/17435/132397291-38566d93-b68d-4ad5-bfda-88384d99f2a1.gif)

Does React handle those refs in a garbage collectible way? or is it just the overhead of React components and rendering?

I made a simple vanilla JS countdown button https://codepen.io/tomalec/pen/VwWPLQy?editors=0011
and it does not have such an effect.

tomalec commented on js/src/components/contact-information/phone-number-card/verification-code-control.js:
```diff
@@ -0,0 +1,136 @@
+	};
+
+	const handleInput = ( e ) => {
+		const { value, dataset } = e.target;
+		const idx = Number( dataset.idx );
+		const digit = value.substr( cursorRef.current, 1 ).replace( /\D/, '' );
+
+		if ( digit !== value ) {
+			e.target.value = digit;
+		}
```
:nail_care: 
We do value to (stringified) number sanitization here, right? a code commend would be nice

eason9487 replied:
Added by https://github.com/woocommerce/google-listings-and-ads/pull/993/commits/fc4ab5c417d1d8c564b0f4bc6a08fcbe79cbb17a.

tomalec commented:
:question: 
I'm not sure if that's an issue here or with my WC5.7, but I got React error when I start typing the first code number:
![type](https://user-images.githubusercontent.com/17435/132412714-61a41856-bfd7-406f-ad85-3b9d5adfa10a.gif)


<details>
<summary>Details</summary>

```
Warning: A component is changing an uncontrolled input of type undefined to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://fb.me/react-controlled-components
    in input (created by Context.Consumer)
    in Input (created by ForwardRef(InputField))
    in ForwardRef(InputField) (created by ForwardRef(InputControl))
    in div (created by Context.Consumer)
    in Container (created by ForwardRef(InputBase))
    in div (created by Context.Consumer)
    in Flex (created by ForwardRef(FlexComponent))
    in ForwardRef(FlexComponent) (created by Context.Consumer)
    in Root (created by ForwardRef(InputBase))
    in ForwardRef(InputBase) (created by ForwardRef(InputControl))
    in ForwardRef(InputControl) (created by ForwardRef)
    in div (created by ForwardRef)
    in ForwardRef (created by VerificationCodeControl)
    in div (created by Context.Consumer)
    in Flex (created by ForwardRef(FlexComponent))
    in ForwardRef(FlexComponent) (created by VerificationCodeControl)
    in VerificationCodeControl (created by VerifyPhoneNumberContent)
    in div (created by Subsection)
    in Subsection (created by VerifyPhoneNumberContent)
    in div (created by Context.Consumer)
    in BodyUI (created by CardBody)
    in CardBody (created by Body)
    in Body (created by VerifyPhoneNumberContent)
    in VerifyPhoneNumberContent (created by DemoVerifyPhoneNumberCard)
    in div (created by Context.Consumer)
    in CardUI (created by Card)
    in Card (created by Card)
    in Card (created by AccountCard)
    in AccountCard (created by DemoVerifyPhoneNumberCard)
    in div (created by DemoVerifyPhoneNumberCard)
    in section (created by DemoVerifyPhoneNumberCard)
    in DemoVerifyPhoneNumberCard (created by Settings)
    in div (created by Settings)
    in Settings (created by ee)
    in Suspense (created by ee)
    in ee (created by bt)
    in div (created by bt)
    in div (created by ht)
    in ht (created by bt)
    in div (created by bt)
    in slot_fill_provider_SlotFillProvider (created by Provider)
    in provider_SlotFillProvider (created by Provider)
    in Provider (created by bt)
    in bt
    in Unknown (created by WithSelect(bt))
    in WithSelect(bt) (created by WithPluginsHydration(WithSelect(bt)))
    in WithPluginsHydration(WithSelect(bt)) (created by Context.Consumer)
    in t
    in t
    in t
    in Unknown (created by WithOptionsHydration(Component))
    in WithOptionsHydration(Component) (created by WithSettingsHydration(WithOptionsHydration(Component)))
    in WithSettingsHydration(WithOptionsHydration(Component)) (created by WithSettingsHydration(WithSettingsHydration(WithOptionsHydration(Component))))
    in WithSettingsHydration(WithSettingsHydration(WithOptionsHydration(Component))) (created by WithCurrentUserHydration(WithSettingsHydration(WithSettingsHydration(WithOptionsHydration(Component)))))
    in WithCurrentUserHydration(WithSettingsHydration(WithSettingsHydration(WithOptionsHydration(Component))))
overrideMethod @ react_devtools_backend.js:4049
printWarning @ react-dom.js?ver=16.13.1:82
error @ react-dom.js?ver=16.13.1:54
updateWrapper @ react-dom.js?ver=16.13.1:1993
updateProperties @ react-dom.js?ver=16.13.1:6383
commitUpdate @ react-dom.js?ver=16.13.1:7676
commitWork @ react-dom.js?ver=16.13.1:20681
commitMutationEffects @ react-dom.js?ver=16.13.1:22911
callCallback @ react-dom.js?ver=16.13.1:182
invokeGuardedCallbackDev @ react-dom.js?ver=16.13.1:231
invokeGuardedCallback @ react-dom.js?ver=16.13.1:286
commitRootImpl @ react-dom.js?ver=16.13.1:22644
unstable_runWithPriority @ react.js?ver=16.13.1:2685
runWithPriority$1 @ react-dom.js?ver=16.13.1:11174
commitRoot @ react-dom.js?ver=16.13.1:22516
finishSyncRender @ react-dom.js?ver=16.13.1:21942
performSyncWorkOnRoot @ react-dom.js?ver=16.13.1:21928
(anonymous) @ react-dom.js?ver=16.13.1:11224
unstable_runWithPriority @ react.js?ver=16.13.1:2685
runWithPriority$1 @ react-dom.js?ver=16.13.1:11174
flushSyncCallbackQueueImpl @ react-dom.js?ver=16.13.1:11219
flushSyncCallbackQueue @ react-dom.js?ver=16.13.1:11207
flushPendingDiscreteUpdates @ react-dom.js?ver=16.13.1:21982
flushDiscreteUpdates @ react-dom.js?ver=16.13.1:21962
finishEventHandler @ react-dom.js?ver=16.13.1:761
batchedEventUpdates @ react-dom.js?ver=16.13.1:795
dispatchEventForLegacyPluginEventSystem @ react-dom.js?ver=16.13.1:3691
attemptToDispatchEvent @ react-dom.js?ver=16.13.1:4390
dispatchEvent @ react-dom.js?ver=16.13.1:4312
unstable_runWithPriority @ react.js?ver=16.13.1:2685
runWithPriority$1 @ react-dom.js?ver=16.13.1:11174
discreteUpdates$1 @ react-dom.js?ver=16.13.1:22022
discreteUpdates @ react-dom.js?ver=16.13.1:803
dispatchDiscreteEvent @ react-dom.js?ver=16.13.1:4291
Show 4 more frames
```

</details>

tomalec commented on js/src/components/contact-information/phone-number-card/verify-phone-number-content.js:
```diff
@@ -0,0 +1,236 @@
+
+		const id = verificationIdRef.current[ method ];
+		verifyPhoneNumber( id, verification.code, method )
+			.then( () => {
+				onPhoneNumberVerified();
+			} )
```
:nail_care: 
Why not just 
```suggestion
			.then( onPhoneNumberVerified )
```

tomalec replied:
:grey_question: BTW, How would that behavior when we re-render the component in the middle of promise resolution?

Wouldn't we call `onPhoneNumberVerified` from the unmounted component, and render the non-verified state?

eason9487 replied:
Since the API calling function `verifyPhoneNumber` is not yet implemented, the additional anonymous function is added to prevent its returned value from being passed as unintended parameters to `onPhoneNumberVerified` via the Promise chain.

> How would that behavior when we re-render the component in the middle of promise resolution?

The countdown timer continues to trigger re-render during the middle of promise resolution if you start verifying before the countdown time up. The component would keep being rendered according to its state and props. 

> Wouldn't we call `onPhoneNumberVerified` from the unmounted component, and render the non-verified state?

Not quite sure your meaning. Even if the `onPhoneNumberVerified` is called from an unmounted component, it should still result in a verified phone number card as below. This part will be added by a subsequent PR.

![image](https://user-images.githubusercontent.com/17420811/132450782-aeca2cf1-60b9-4edf-8e81-2dd366990572.png)

tomalec replied:
> Not quite sure your meaning. Even if the onPhoneNumberVerified is called from an unmounted component, it should still result in a verified phone number card as below. This part will be added by a subsequent PR.

What I mean is, more-or-less:
1. Call `VerifyPhoneNumberContent( { onPhoneNumberVerified: A } )`
2. Click "Verify…" (`VerifyPhoneNumberContent~handleVerifyClick`=>`verifyPhoneNumber()`)
3. Call `VerifyPhoneNumberContent( { onPhoneNumberVerified: B } )`
4. Promise from 2. resolves, calling `A()`

eason9487 replied:
Do you mean something like this https://codesandbox.io/s/sad-khayyam-5zpus?file=/src/App.js ?

1. Click the button to chose Buzz or Woody - set the `A` callback
2. Click "Greet!"
3. Click the button opposite to the one picked from 1. - set the `B` callback
4. The resolved Promise from 2. calls `A()`
5. The welcomed person and the actual greeted person might be different. Is this your concern?
    ![Kapture 2021-09-08 at 20 12 06](https://user-images.githubusercontent.com/17420811/132507130-097a3704-7915-40ad-a6fb-9040517fc3d9.gif)

tomalec replied:
> `fetchSomebo` 

:rofl: :heart_eyes: 

-----


Yes, something like this. but, in our case when we

tomalec replied:
…in our case, that's `<Greeting>` button that is about to be re-rendered during promise resolution: https://codesandbox.io/s/nifty-pateu-2zi5q?file=/src/App.js

Try:
1. Click "Buzz…"
2. Click "Greet!"
3. You should see "Verifying…" as expected
4. Before it's resolved, click "Sherif …"
5. Click "Greet!" again.
6. You should see "Verifying…" as expected, before "Howdy, Buzz …" is rendered.
7. :red_square:  but then "Verifying…" is gone, even though, there's still verifying of "Sherif" being done in the background.

![verifying](https://user-images.githubusercontent.com/17435/132677680-c48c1de3-5009-4870-8201-6451764468cd.gif)

tomalec replied:
I think I managed to reproduce my hypothetical problem at https://github.com/woocommerce/google-listings-and-ads/issues/1004

eason9487 replied:
The problem in 7. would become a concern when there's a way to trigger another same task of promise during the previous one is not yet fulfilled, such as step 5. But this case should not happen in the `<VerifyPhoneNumberContent>`.

And issue #1004 is more about updating state on an unmounted component, like the `verifying`, `error`, and countdown states within the `<VerifyPhoneNumberContent>`. It's an issue but I think it doesn't relate to the verified state. The verified state will be handled in the upper scope or the wp-data store. Even though a user navigates away during an issued promise process, these unexpected cases could be handled by the consumer component of `<VerifyPhoneNumberContent>` when the `onPhoneNumberVerified` calling back.

tomalec reviewed (commented):
Tested locally and reviewed the code. The functionality works as expected. Code looks good, left few cosmetic (:nail_care:) comment.
However, I got few doubts marked with :question:: https://github.com/woocommerce/google-listings-and-ads/pull/993#issuecomment-914638887 https://github.com/woocommerce/google-listings-and-ads/pull/993#issuecomment-914551267

eason9487 committed:
Fix English grammar.

Address https://github.com/woocommerce/google-listings-and-ads/pull/993#discussion_r703605349

Co-authored-by: Tomek Wytrębowicz <tomalecpub@gmail.com>

eason9487 committed:
Change the naming and structure for constants of verification methods

Address https://github.com/woocommerce/google-listings-and-ads/pull/993#discussion_r703565486

eason9487 committed:
Add a comment for the sanitized digit in the VerificationCodeControl

Address https://github.com/woocommerce/google-listings-and-ads/pull/993#discussion_r703831236

eason9487 committed:
Prevent the warning of uncontrolled input

Address https://github.com/woocommerce/google-listings-and-ads/pull/993#issuecomment-914638887

eason9487 committed:
Disable the resend button of verification code when verifying phone number

eason9487 commented:
@tomalec, thanks for catching this! 🙏 

> I'm not sure if that's an issue here or with my WC5.7, but I got React error when I start typing the first code number

Just realized that the `<InputControl>` always specifies `onChange` so that the input is forced to be a controlled type. Seems that there‘s no way to specify `<InputControl>` as the uncontrolled type, so I specify `value` prop to avoid the warning. Fixed by https://github.com/woocommerce/google-listings-and-ads/pull/993/commits/808e70ec309ce9e1f4a80b1de560a39ea2af0cc2

p.s. I saw the autocomplete selection shows up from your GIF, I added `autoComplete="off"` to prevent it from appearing.

eason9487 commented:
@tomalec,

> I noticed that the countdown takes ~1MB of memory, which looks quite expensive for such a simple component.

> Does React handle those refs in a garbage collectible way? or is it just the overhead of React components and rendering?

Haven't used the memory profiling tool before. Hope my understanding of it is correct. I recorded two profiles by the "Allocation instrumentation on timeline" mode. The videos and an exported heaptimeline file are attached below.

### Memory allocation at the beginning, end, and during the countdown

https://user-images.githubusercontent.com/17420811/132480290-2764154a-a7e3-403a-822a-473d3da47a93.mp4

[Heap-20210908T153250.heaptimeline.zip](https://github.com/woocommerce/google-listings-and-ads/files/7127654/Heap-20210908T153250.heaptimeline.zip) could be loaded into the memory profiling tool.

(To avoid some noise from the sibling component, I removed `<VerificationCodeControl>` that trigger `focus()` before recording)

From the record, the heaps that occurred on timeline are:
- ≈ 2s: click on resend button, it recorded some memory used to create the countdown related tasks.
- every second between 2s ~ 1m1s, it recorded a spike due to the remaining seconds updated. But they didn't leave uncollected memory.
- ≈ 26s: recorded a heartbeat event issued from WP.
- ≈ 47s: recorded a heap that an AJAX request issued to `/wp-admin/admin-ajax.php`, and another heap for handling the response.
- after 1m2s: there're two heaps for re-rendering the `<VerifyPhoneNumberContent>` and create the butter after the countdown ends.

### Memory allocation for the countdown back to back

https://user-images.githubusercontent.com/17420811/132490669-4a785e49-0d34-41ca-9746-13d088d6f64f.mp4

From the record, you can see the allocated heap of the countdown end will be GC after starting a new countdown.

---

From the first video, the allocated memory is 44.1 kB, and the more close size would be 32.3 kB if deduct the 11.8 kB used by the AJAX request. I think the memory usage should be ok and no obvious memory leak here. Not sure why it increased ~ 1 MB on your end. 🤔

tomalec commented:
Thanks for the detailed investigation.
I run the profiler on my local and without `VerificationCodeControl` and the heap is cleared completely.

But when I run it with `VerificationCodeControl` there are some leftovers
![mem leak](https://user-images.githubusercontent.com/17435/132549509-93a70479-9f50-4c98-9d8a-fcc5074d30c5.gif)


When I switch between SMS and phone, there are reduced but still there.
![switching tosms phone](https://user-images.githubusercontent.com/17435/132549519-276d32ab-3595-4add-a9f8-8da326d0431b.gif)


However, after the timer is done/reset and we switch between SMS/phone it is not so significant in size.

I think we can ignore it for now, or leave for later 📜

eason9487 committed:
Revert "Add a temporary component to demonstrate the phone number verification card."

This reverts commit d86dde759dda1909d0191a2da6c14a1d6c709748.

tomalec commented:
Sorry for being late, but I discovered two more issues: https://github.com/woocommerce/google-listings-and-ads/issues/1002 https://github.com/woocommerce/google-listings-and-ads/issues/1003

eason9487 commented:
> Sorry for being late, but I discovered two more issues: #1002 #1003

@tomalec, thanks for catching these bugs. 👍  I had addressed them on my local dev env. Will open PR tomorrow.
